### PR TITLE
Use testcontainers for integration tests

### PR DESCRIPTION
**Pull request by Barsonax**
## Breaking change

This will break the existing integration build.

## Proposed change

Use the testcontaiers library to automatically launch a homeassistant docker container in order to run integration tests. Except for having docker installed the user does not have to setup anything manually.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainer.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            password = _configuration.Password,
+            language = "en-GB"
+        }));
+
+        onboardingResult.EnsureSuccessStatusCode();
+
+        return (await onboardingResult.Content.ReadFromJsonAsync<HomeAssistantAuthorizeResult>())!;
+    }
+
+    public async Task<HomeAssistantTokenResult> GenerateApiToken(string authCode)
```
Comment (Barsonax) [inline tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainer.cs:44]: This api token is valid for 30 minutes. Probably more than enough for the time being. 

If we want more then we need to get a long lived access token using websockets.

File: tests/Integration/NetDaemon.Tests.Integration/NetDaemon.Tests.Integration.csproj [hunk 2]
```diff
@@ -11,8 +11,8 @@
         <PackageReference Include="FluentAssertions" Version="6.11.0" />
         <PackageReference Include="Microsoft.Extensions.Hosting" Version="7.0.1" />
         <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.6.2" />
+        <PackageReference Include="Testcontainers" Version="3.3.0" />
         <PackageReference Include="xunit" Version="2.4.2" />
-        <PackageReference Include="Xunit.DependencyInjection" Version="8.7.1" />
```
Comment (Barsonax) [inline tests/Integration/NetDaemon.Tests.Integration/NetDaemon.Tests.Integration.csproj:15]: I removed this as I needed more control over the lifetimes. Still going to check if Xunit.DependencyInjection has some way of providing that control so I don't have to remove it.

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantLifetime.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+﻿using System.IO;
+using System.Threading.Tasks;
+using NetDaemon.Tests.Integration.Helpers.HomeAssistantTestContainer;
+using Xunit;
+
+namespace NetDaemon.Tests.Integration.Helpers;
+
+public class HomeAssistantLifetime : IAsyncLifetime
+{
+    private readonly HomeAssistantContainer _homeassistant = new HomeAssistantContainerBuilder()
```
Comment (Barsonax) [inline tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantLifetime.cs:10]: I still have to make it possible to specify the version here from the CI so we can test against multiple versions.

File: tests/Integration/NetDaemon.Tests.Integration/NetDaemon.Tests.Integration.csproj (hunk 2 above)
Comment (Barsonax) [reply tests/Integration/NetDaemon.Tests.Integration/NetDaemon.Tests.Integration.csproj:15]: There does not seem to be a way so iam leaving this out

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainer.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+using System.Threading.Tasks;
+using DotNet.Testcontainers.Containers;
+using Microsoft.Extensions.Logging;
+
+namespace NetDaemon.Tests.Integration.Helpers.HomeAssistantTestContainer;
+
+public class HomeAssistantContainer : DockerContainer
+{
+    private readonly HomeAssistantConfiguration _configuration;
+    public ushort Port => GetMappedPublicPort(8123);
```
Comment (helto4real) [inline tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainer.cs:15]: I am guessing this code makes sure the container has a random available port that maps to the internam 8123. That is great for CI! You never know what ports ar available!

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainer.cs (hunk 1 above)
Comment (helto4real) [reply tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainer.cs:44]: That is enough indeed. If not we have other problems :)

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainerBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    public HomeAssistantContainerBuilder(HomeAssistantConfiguration dockerResourceConfiguration) : base(dockerResourceConfiguration)
+    {
+        DockerResourceConfiguration = dockerResourceConfiguration;
+    }
+
+    protected override HomeAssistantContainerBuilder Init()
+    {
+        return base.Init()
+            .WithImage("homeassistant/home-assistant:stable")
```
Comment (helto4real) [inline tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainerBuilder.cs:27]: Interesting the choose "stable" and not latest. Seems like a good practice

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantLifetime.cs (hunk 3 above)
Comment (helto4real) [reply tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantLifetime.cs:11]: You do not think the environment fixes that. Or do you mean you need to fix the CI too?

**Review by helto4real — COMMENTED** [review]
I know you spent a lot of time getting this to work but the end-result seem really simple. We can build from this, adding more tests etc. Great job. Will approve as soon as the CI jobs are updated.

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantLifetime.cs (hunk 3 above)
Comment (Barsonax) [reply tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantLifetime.cs:11]: I made this comment before I added the environment variable

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainer.cs (hunk 4 above)
Comment (Barsonax) [reply tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainer.cs:15]: Yup and it seems to be doing a good job at selecting a port that is actually available.

File: tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainerBuilder.cs (hunk 5 above)
Comment (Barsonax) [reply tests/Integration/NetDaemon.Tests.Integration/Helpers/HomeAssistantTestContainer/HomeAssistantContainerBuilder.cs]: I did not really think about this and just went for stable as it seemed the most obvious choice.

**Review by helto4real — APPROVED** [review]
Looks great! Thanks alot for this!

**Pull request merged**